crates/bevy_ecs/src/bundle/info.rs

```diff
@@ -543,12 +624,96 @@ impl Bundles {
             });
         *bundle_id
     }
+
+    /// Checks if the bundles containing `requiree` can have their required components be updated,
+    /// in which case this returns `Ok(true)`.
+    ///
+    /// Returns `Ok(false)` if there are no known bundles with this component.
+    ///
+    /// Returns the [`RequiredComponentsError::RemovedFromArchetype`] error if bundles cannot be updated
+    /// in which case the registration of any new required component must be refused.
+    pub(crate) fn verify_to_refresh_required_components(
+        &self,
+        requiree: ComponentId,
+    ) -> Result<bool, RequiredComponentsError> {
+        let Some(bundles_with_requiree) = self
+            .components_in_bundles
+            .get(requiree.index())
+            .filter(|bundles| !bundles.is_empty())
+        else {
+            // no bundle with `requiree` exists
+            return Ok(false);
+        };
+
+        // `EntityWorldMut::remove_with_requires` generate edges between archetypes where the required
+        // components matter for the the target archetype. These bundles cannot receive new required
+        // components as that would invalidate these edges.
+        // The mechanism is using `Self::contributed_bundle_ids` to track the bundles so we check here if it
+        // has any overlapping `BundleId`s with `bundles_with_requiree`.
+        // TODO: Remove this error and update archetype edges accordingly when required components are added
+        if bundles_with_requiree.iter().any(|bundles_with_requiree| {
+            self.contributed_bundle_ids
+                .values()
+                .any(|bundle: &BundleId| bundles_with_requiree == bundle)
+        }) {
+            return Err(RequiredComponentsError::RemovedFromArchetype(requiree));
```

Comment:
The error message doesn't look appropriate. `contributed_bundle_ids` can be populated when inserting too, so you might get this error when a component was never removed from an archetype.